### PR TITLE
Js toggles

### PR DESCRIPTION
**Pull request by dannyroberts**
@dimagi/js-team what do you think?

:blowfish:

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 8 errors:

* corehq/apps/hqwebapp/static/hqwebapp/js/toggles.js, line 7 - Bad line breaking before '+'.
* corehq/apps/hqwebapp/static/hqwebapp/js/toggles.js, line 13 - Missing semicolon.
* corehq/apps/hqwebapp/static/hqwebapp/js/toggles.js, line 16 - Missing semicolon.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a string and instead saw %.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected ':' and instead saw 'load'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a JSON value.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '}' and instead saw 'hq_shared_tags'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '(end)' and instead saw '%'.

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a string and instead saw %.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected ':' and instead saw 'load'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a JSON value.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '}' and instead saw 'hq_shared_tags'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '(end)' and instead saw '%'.

**Comment by dannyroberts** [comment]
(not yet complete, I still need to add cache busting.)

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a string and instead saw %.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected ':' and instead saw 'load'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a JSON value.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '}' and instead saw 'hq_shared_tags'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '(end)' and instead saw '%'.

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a string and instead saw %.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected ':' and instead saw 'load'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a JSON value.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '}' and instead saw 'hq_shared_tags'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '(end)' and instead saw '%'.

**Comment by dannyroberts** [comment]
Ok, I've tested it enough locally (both just js change itself and the caching/cachebusting) that it's ready to go in my view. Let me know what you think!

File: corehq/apps/hqwebapp/templatetags/hq_shared_tags.py [hunk 1]
```diff
@@ -228,6 +227,29 @@ def toggle_enabled(request, toggle_or_toggle_name):
...
+        domain_cb=toggle_js_domain_cachebuster(domain),
+        user_cb=toggle_js_user_cachebuster(username),
+    )
+
+
+@quickcache(['domain'], timeout=30 * 24 * 60 * 60)
+def toggle_js_domain_cachebuster(domain):
+    # to get fresh cachebusters on the next deploy
+    # change the date below (output from *nix `date` command)
+    #   Thu Mar  3 16:21:30 EST 2016
```
Comment (millerdev) [inline corehq/apps/hqwebapp/templatetags/hq_shared_tags.py]: I don't understand this comment. Does it mean changing date in this comment will cause the caches to be busted on the next deploy? How frequently do you expect that we'll want to do that?

File: corehq/apps/hqwebapp/urls.py [hunk 2]
```diff
@@ -33,9 +33,10 @@
 domain_specific = patterns('corehq.apps.hqwebapp.views',
     url(r'^$', 'redirect_to_default', name='domain_homepage'),
     url(r'^login/$', 'domain_login', name='domain_login'),
-    url(r'^login/mobile/$', 'domain_login', name='domain_mobile_login', 
+    url(r'^login/mobile/$', 'domain_login', name='domain_mobile_login',
         kwargs={'template_name': 'login_and_password/mobile_login.html'}),
-    url(r'^retreive_download/(?P<download_id>[0-9a-fA-Z]{25,32})/$', 
-        'retrieve_download', {'template': 'style/includes/file_download.html' },
-        name='hq_soil_download')
+    url(r'^retreive_download/(?P<download_id>[0-9a-fA-Z]{25,32})/$',
```
Comment (millerdev) [inline corehq/apps/hqwebapp/urls.py]: I think it probably doesn't matter (and I it looks like you didn't change it), but `retreive` is spelled incorrectly (it's spelled correctly on the next line)

File: corehq/apps/toggle_ui/views.py [hunk 3]
```diff
@@ -157,12 +159,17 @@ def page_context(self):
...
         for entry in changed_entries:
             if entry.startswith(NAMESPACE_DOMAIN):
                 domain = entry.split(":")[-1]
-                self.static_toggle.save_fn(domain, entry in currently_enabled)
+                if self.static_toggle.save_fn is not None:
+                    self.static_toggle.save_fn(domain, entry in currently_enabled)
+                toggle_js_domain_cachebuster.clear(domain)
+            else:
+                # these are sent down with no namespace
+                assert ':' not in entry
```
Comment (millerdev) [inline corehq/apps/toggle_ui/views.py]: might want to change this to `assert ':' not in entry, entry` to give more debugging context if this ever fails

File: corehq/apps/hqwebapp/templatetags/hq_shared_tags.py (hunk 1 above)
Comment (dannyroberts) [inline corehq/apps/hqwebapp/templatetags/hq_shared_tags.py]: We'll want that if we change the format of that file, say, to also include privileges or something. We'll then need all browsers to reload the file. Changing the date in this function changes the quickcache keys, which will result in generating a new value here, which will result in the browser doing a fresh request.

File: corehq/toggles.py [hunk 4]
```diff
@@ -127,22 +127,44 @@ def all_toggles():
...
 
 
+def toggle_values_by_name(username=None, domain=None):
+    """
+    Loads all toggles into a dictionary for use in JS
+
+    all toggles (including those not enabled) are included
+    """
+    return {toggle_name: (toggle.enabled(username) or toggle.enabled(domain))
+            for toggle_name, toggle in all_toggles_by_name().items()}
```
Comment (millerdev) [inline corehq/toggles.py]: does this hit the database twice per toggle (once for `username` and once for `domain`)? If yes, would it make sense to refactor it to issue a single query that gets all toggles and their enabled/disabled values? (is that even possible?)

File: corehq/apps/hqwebapp/templatetags/hq_shared_tags.py (hunk 1 above)
Comment (millerdev) [inline corehq/apps/hqwebapp/templatetags/hq_shared_tags.py]: Oh, interesting. I may be missing something, but does that mean that any time this file is changed all toggle caches will be busted? Do we want that?

Comment (dannyroberts) [inline corehq/apps/hqwebapp/templatetags/hq_shared_tags.py]: It'll only change when the code of this particular function (or the other cachebuster function) changes. And I think we more or less do want that. Anytime we're fiddling with how cachebusting works, we'll probably want everyone to be on whatever the new system is—and recomputing everyone's toggles json is not an extraordinarily intensive computation.

Comment (millerdev) [inline corehq/apps/hqwebapp/templatetags/hq_shared_tags.py]: I didn't know we had individual function source change detection/cache busting, but that's cool. :+1:

File: corehq/toggles.py (hunk 4 above)
Comment (dannyroberts) [inline corehq/toggles.py]: Toggles come with some built-in caching, so I think this is fine. If the particular kind of caching it does turns out to be insufficient, there's a pretty easy caching solution to this particular problem. My attitude here is to write the code the way you'd want to write it, and then if performance is a problem fix the way the interface is cached. In this particular case I believe the caching is sufficient, and also the amount data and the frequency with which it'll be requested are both low enough that I'm not too worried in the short term.

Comment (millerdev) [inline corehq/toggles.py]: :+1:

File: corehq/apps/hqwebapp/templatetags/hq_shared_tags.py (hunk 1 above)
Comment (dannyroberts) [inline corehq/apps/hqwebapp/templatetags/hq_shared_tags.py]: It *is* pretty cool! (If I may say so myself!) Source: https://github.com/dimagi/commcare-hq/blob/12a0ac725b1e0d7dc325d1e6411c23a7a3ca22a2/corehq/util/quickcache.py#L68 and comment: https://github.com/dimagi/commcare-hq/blob/12a0ac725b1e0d7dc325d1e6411c23a7a3ca22a2/corehq/util/quickcache.py#L256

**Comment by benrudolph** [comment]
nice danny! this is sweet :zap: :traffic_light:

File: corehq/apps/toggle_ui/views.py (hunk 3 above)
Comment (dannyroberts) [inline corehq/apps/toggle_ui/views.py]: ok I'll do that, thanks!

File: corehq/apps/hqwebapp/urls.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/hqwebapp/urls.py]: ha! I think I'm gonna leave it...

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a string and instead saw %.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected ':' and instead saw 'load'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected a JSON value.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '}' and instead saw 'hq_shared_tags'.
* corehq/apps/hqwebapp/templates/hqwebapp/js/toggles_template.js, line 1 - Expected '(end)' and instead saw '%'.

**Comment by dannyroberts** [comment]
:white_check_mark: (can ignore last commit's tests, very insubstantive change)

feedback so far addressed, let me know if you have anything else, but I think it's good to merge!

**Pull request merged**